torchtitan/experiments/deepseek_v3/moe_kernels.py

```diff
@@ -0,0 +1,261 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+import triton
+import triton.language as tl
+
+
+__all__ = ["generate_permute_indices"]
+
+
+# parallelized kernel
+@triton.jit
+def _fill_indices_kernel(
+    tokens_per_expert_group_ptr,
+    start_index_values_ptr,
+    write_offsets_ptr,
+    output_ptr,
+    experts_per_rank: tl.constexpr,
+    num_ranks: tl.constexpr,
+    total_experts: tl.constexpr,
+    max_blocks: tl.constexpr,
+    experts_per_block: tl.constexpr,
+    BLOCK_SIZE: tl.constexpr,  # Number of threads per block
+):
+    pid = tl.program_id(axis=0)
+
+    # we loop if needed
+    start_expert = pid * experts_per_block
+    end_expert = tl.minimum(start_expert + experts_per_block, total_experts)
+
+    for expert_id in range(start_expert, end_expert):
+        # only process if valid expert
+        if expert_id < experts_per_rank:
```

Comment:
pid = tl.program_id(axis=0)  # We use a 1D launch grid so axis is 0.
    # The total number of programs in the launch grid.
    num_programs = tl.num_programs(axis=0)
    # We map the programs (blocks) to the experts.
    for expert_id in tl.range(pid, experts_per_rank, step=num_programs):

Reply:
we should loop over local experts instead of total_experts.

Reply:
got it, thanks for catching.  Updated the looping!